File: src/engine/CalculatorEngine.test.ts
/**
 * CalculatorEngine.test.ts
 * ============================================================
 * JUnit-style unit test suite for the calculator engine.
 * Android equivalent: app/src/test/java/.../CalculatorEngineTest.kt
 * running on the JVM via JUnit 4. Here we mirror it with a tiny
 * assertion harness executed in the browser console at startup.
 * ============================================================
 */
import {
  calculateExpression,
  formatResult,
  tokenize,
} from "./CalculatorEngine";

const results: { name: string; pass: boolean; detail?: string }[] = [];
let current = "";

function test(name: string, fn: () => void): void {
  current = name;
  try {
    fn();
    results.push({ name, pass: true });
  } catch (e) {
    results.push({ name, pass: false, detail: e instanceof Error ? e.message : String(e) });
  }
}

function assertApprox(actual: number | null | undefined, expected: number, epsilon = 1e-9): void {
  if (actual === null || actual === undefined || Math.abs(actual - expected) > epsilon) {
    throw new Error(`${current}: expected ≈${expected}, got ${actual}`);
  }
}

function assertError(expr: string, messagePart: string): void {
  const res = calculateExpression(expr);
  if (res.ok) throw new Error(`${current}: "${expr}" should fail, but returned ${res.result.value}`);
  if (!res.error.includes(messagePart)) {
    throw new Error(`${current}: "${expr}" error was "${res.error}", expected to include "${messagePart}"`);
  }
}

/* ------------------- test suite ------------------- */
test("addition of two integers", () => {
  const r = calculateExpression("2+3");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 5);
});

test("operator precedence (× binds tighter than +)", () => {
  const r = calculateExpression("2+3×4");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 14);
});

test("parentheses override precedence", () => {
  const r = calculateExpression("(2+3)×4");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 20);
});

test("division by zero returns error", () => {
  assertError("5÷0", "Cannot divide by zero");
  assertError("1÷(2−2)", "Cannot divide by zero");
});

test("negative numbers via unary minus", () => {
  const r = calculateExpression("−5×3");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, -15);
});

test("trigonometric functions use degrees", () => {
  const sin = calculateExpression("sin(30)");
  const cos = calculateExpression("cos(60)");
  const tan = calculateExpression("tan(45)");
  if (!sin.ok || !cos.ok || !tan.ok) throw new Error("trig failed");
  assertApprox(sin.result.value, 0.5);
  assertApprox(cos.result.value, 0.5);
  assertApprox(tan.result.value, 1);
});

test("tan(90) is undefined", () => {
  assertError("tan(90)", "undefined");
});

test("logarithms", () => {
  const log10 = calculateExpression("log(1000)");
  const ln = calculateExpression("ln(e)");
  if (!log10.ok || !ln.ok) throw new Error("log failed");
  assertApprox(log10.result.value, 3);
  assertApprox(ln.result.value, 1);
});

test("exponentials", () => {
  const sq = calculateExpression("3^4");
  const cube = calculateExpression("2^10");
  if (!sq.ok || !cube.ok) throw new Error("pow failed");
  assertApprox(sq.result.value, 81);
  assertApprox(cube.result.value, 1024);
});

test("factorial", () => {
  const f = calculateExpression("5!");
  if (!f.ok) throw new Error(f.error);
  assertApprox(f.result.value, 120);
});

test("factorial of negative errors", () => {
  assertError("(−3)!", "negative");
});

test("square root", () => {
  const r = calculateExpression("sqrt(144)");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 12);
});

test("square root of negative errors", () => {
  assertError("sqrt(−4)", "negative");
});

test("percentage", () => {
  const r = calculateExpression("25%×200");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 50);
});

test("large number formatting switches to scientific", () => {
  const r = calculateExpression("123456789012345678");
  if (!r.ok) throw new Error(r.error);
  const formatted = formatResult(r.result.value);
  if (!formatted.includes("e")) throw new Error(`expected scientific notation, got ${formatted}`);
});

test("very small numbers use scientific notation", () => {
  const r = calculateExpression("1÷10^12");
  if (!r.ok) throw new Error(r.error);
  const formatted = formatResult(r.result.value);
  if (!formatted.includes("e")) throw new Error(`expected scientific notation, got ${formatted}`);
});

test("implicit multiplication 2(3+4)", () => {
  const r = calculateExpression("2(3+4)");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 14);
});

test("implicit multiplication before functions: 2sin(30)", () => {
  const r = calculateExpression("2sin(30)");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 1);
});

test("factorial then implicit multiply: 2!sin(30)", () => {
  const r = calculateExpression("2!sin(30)");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 1);
});

test("power after unary minus: 2^−3", () => {
  const r = calculateExpression("2^−3");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 0.125);
});

test("deep parentheses and unary chains", () => {
  const r = calculateExpression("((2+3)×(4−1))^2");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 225);
});

test("tokenizer rejects unknown characters", () => {
  const t = tokenize("2 & 3");
  if (typeof t !== "string") throw new Error("expected tokenizer error");
});

test("empty expression errors", () => {
  assertError("", "Empty expression");
});

test("right-associative power: 2^3^2 = 512", () => {
  const r = calculateExpression("2^3^2");
  if (!r.ok) throw new Error(r.error);
  assertApprox(r.result.value, 512);
});

/* ------------------- report ------------------- */
const failed = results.filter((r) => !r.pass);

if (typeof window !== "undefined" && !(window as unknown as { __calcTestsRun?: boolean }).__calcTestsRun) {
  (window as unknown as { __calcTestsRun?: boolean }).__calcTestsRun = true;
  if (failed.length) {
    // eslint-disable-next-line no-console
    console.error(`[NeonCalc] engine tests: ${results.length - failed.length}/${results.length} passed`, failed);
  } else {
    // eslint-disable-next-line no-console
    console.info(`[NeonCalc] engine tests: all ${results.length} passed ✅`);
  }
}

export { results as testResults, assertApprox };

File: android/app/src/test/java/com/neoncalc/app/engine/CalculatorEngineTest.kt
package com.neoncalc.app.engine

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * CalculatorEngineTest — JUnit 4 suite for the math core.
 * Covers arithmetic, precedence, parentheses, errors (division by zero),
 * trigonometry, logs, powers, factorial, percentages, large/small number
 * formatting and implicit multiplication.
 */
class CalculatorEngineTest {

    private fun value(expr: String): Double {
        val res = CalculatorEngine.calculate(expr)
        val failure = (res as? CalculatorEngine.EvalResult.Failure)?.error
        assertTrue("Expected success for '$expr' but got: $failure", res is CalculatorEngine.EvalResult.Success)
        return (res as CalculatorEngine.EvalResult.Success).value
    }

    private fun assertError(expr: String, part: String) {
        val res = CalculatorEngine.calculate(expr)
        assertTrue("Expected failure for '$expr'", res is CalculatorEngine.EvalResult.Failure)
        val msg = (res as CalculatorEngine.EvalResult.Failure).error
        assertTrue("Error '$msg' should contain '$part'", msg.contains(part))
    }

    /* ---------- basic arithmetic ---------- */
    @Test fun addition() = assertEquals(5.0, value("2+3"), 1e-9)

    @Test fun multiplication() = assertEquals(56.0, value("7×8"), 1e-9)

    @Test fun division() = assertEquals(2.25, value("9÷4"), 1e-9)

    @Test fun operatorPrecedence() = assertEquals(14.0, value("2+3×4"), 1e-9)

    @Test fun parenthesesOverridePrecedence() = assertEquals(20.0, value("(2+3)×4"), 1e-9)

    @Test fun subtractionPrecedence() = assertEquals(4.0, value("10−3×2"), 1e-9)

    @Test fun decimals() = assertEquals(0.3, value("0.1+0.2"), 1e-9)

    /* ---------- errors ---------- */
    @Test fun divisionByZero() {
        assertError("5÷0", "Cannot divide by zero")
        assertError("1÷(2−2)", "Cannot divide by zero")
    }

    @Test fun emptyExpressionErrors() = assertError("", "Empty expression")

    @Test fun unknownCharacterErrors() = assertError("2 & 3", "Unsupported character")

    /* ---------- negatives ---------- */
    @Test fun unaryMinus() = assertEquals(-15.0, value("−5×3"), 1e-9)

    @Test fun negativeAfterPower() = assertEquals(0.125, value("2^−3"), 1e-9)

    /* ---------- scientific functions ---------- */
    @Test fun sineDegrees() = assertEquals(0.5, value("sin(30)"), 1e-9)

    @Test fun cosineDegrees() = assertEquals(0.5, value("cos(60)"), 1e-9)

    @Test fun tangentDegrees() = assertEquals(1.0, value("tan(45)"), 1e-9)

    @Test fun tan90IsUndefined() = assertError("tan(90)", "undefined")

    @Test fun logBase10() = assertEquals(3.0, value("log(1000)"), 1e-9)

    @Test fun naturalLogOfE() = assertEquals(1.0, value("ln(e)"), 1e-9)

    @Test fun squareRoot() = assertEquals(12.0, value("sqrt(144)"), 1e-9)

    @Test fun sqrtOfNegativeErrors() = assertError("sqrt(−4)", "negative")

    @Test fun factorial() = assertEquals(120.0, value("5!"), 1e-9)

    @Test fun factorialOfNegativeErrors() = assertError("(−3)!", "negative")

    @Test fun percentage() = assertEquals(50.0, value("25%×200"), 1e-9)

    /* ---------- powers + implicit multiplication ---------- */
    @Test fun power() = assertEquals(1024.0, value("2^10"), 1e-9)

    @Test fun rightAssociativePower() = assertEquals(512.0, value("2^3^2"), 1e-9)

    @Test fun implicitMultiplyParens() = assertEquals(14.0, value("2(3+4)"), 1e-9)

    @Test fun implicitMultiplyFunction() = assertEquals(1.0, value("2sin(30)"), 1e-9)

    @Test fun nestedExpression() = assertEquals(225.0, value("((2+3)×(4−1))^2"), 1e-9)

    /* ---------- formatting: large & small numbers ---------- */
    @Test fun largeNumbersUseScientificNotation() {
        val formatted = CalculatorEngine.formatResult(value("123456789012345678"))
        assertTrue("Expected scientific notation, got $formatted", formatted.contains("e"))
    }

    @Test fun verySmallNumbersUseScientificNotation() {
        val formatted = CalculatorEngine.formatResult(value("1÷10^12"))
        assertTrue("Expected scientific notation, got $formatted", formatted.contains("e"))
    }

    @Test fun trailingZerosAreTrimmed() {
        assertEquals("5", CalculatorEngine.formatResult(5.0))
        assertEquals("0.25", CalculatorEngine.formatResult(0.25))
    }
}
